충남대 BE_좌진우 2단계 - 회원 로그인

jinwoojwa opened the pull request:
## Step 2 - 회원 로그인

### 2단계 미션 구현 

✅ API 명세 

| URL                  | 메서드  | 기능    | 설명           |
|----------------------|------|-------|--------------|
| /api/members/register        | POST | 회원 가입 | 이메일과 비밀번호로 가입, 성공 시 토큰 반환      |
| /api/members/login | POST  | 로그인 | 이메일과 비밀번호로 로그인, 성공 시 토큰 반환  |

✅ 🔐 JWT 토큰 처리

- `JJWT` 라이브러리를 이용하여 토큰을 발급했습니다.
- `JwtUtil`라는 이름의 유틸 클래스에서 토큰 생성 로직을 수행하도록 구현했습니다.
- 추후 다른 메서드들을 추가하여 인증 필터 적용에 활용할 수 있도록 할 예정입니다.

✅ ⚠️ 예외 처리

- 커스텀 예외 클래스를 추가하고, `GlobalExceptionHandler`에서 예외를 처리하도록 했습니다.
    + `DuplicateEmailException`: 이미 존재하는 이메일
    + `LoginFailedException`: 잘못된 이메일 또는 비밀번호

✅ 추가 구현 계획 및 보충 사항

- 현재 로그인, 회원 가입 기능만 추가하였고, 요청마다 Authorization 헤더를 확인해서 인증을 수행하지는 않습니다.
- 따라서, 요청마다 인증을 수행하기 위한 `필터`를 추가할 계획입니다.
- 회원을 조회,추가 등을 할 수 있는 관리자 화면도 구현 예정입니다.
- 또한 현재 비밀번호가 평문으로 사용되고 있는데, 비밀번호 암호화도 수행할 예정입니다.
- 이에 따른 테스트 코드도 작성 예정입니다.

### 💻 검증 (postman)

- 회원 가입

![image](https://github.com/user-attachments/assets/4971452c-42de-465b-b107-8c24c1940532)
![image](https://github.com/user-attachments/assets/fd4a826c-2143-4be1-b350-2c4d57eba199)
![image](https://github.com/user-attachments/assets/8805ec28-5260-47f6-9915-8cb20f06e67e)

- 로그인

![image](https://github.com/user-attachments/assets/ae71e136-644e-43b6-bf69-43884f0520f2)
![image](https://github.com/user-attachments/assets/ba68087d-9684-47a4-a1f0-24ddc6c81bf3)

jinwoojwa committed:
docs(README): 기능 요구 사항 정리

jinwoojwa committed:
feat: member 테이블 DDL 및 초기 데이터 추가

jinwoojwa committed:
chore: JJWT 라이브러리 의존성 추가

jinwoojwa committed:
feat: JWT 기반 회원가입/로그인 로직 구현

jinwoojwa committed:
feat: 회원 인증 관련 커스텀 예외 및 전역 예외 처리 추가

woodying commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,35 @@
+    @PostMapping("/register")
+    public ResponseEntity<MemberResponseDto> signUp(@RequestBody @Valid MemberRequestDto requestDto) {
+        MemberResponseDto responseDto = authMemberService.signUp(requestDto);
+        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto);
```
201 코드를 잘 활용하고 있네요 👍👍

woodying commented on src/main/java/gift/dto/MemberRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+import jakarta.validation.constraints.Size;
+
+public record MemberRequestDto(
+        @NotBlank(message = "이메일을 입력해주세요.")
+        @Email(message = "유효한 이메일 형식이 아닙니다.")
+        String email,
+
+        @NotBlank(message = "비밀번호를 입력해주세요.")
+        @Size(min = 8, max = 20, message = "비밀번호는 8자 이상 20자 이하여야 합니다.")
+        String password,
+        String role // USER, ADMIN
```
유효성 검증 좋습니다 👏

woodying commented on src/main/java/gift/entity/Member.java:
```diff
@@ -0,0 +1,25 @@
+    private Long id;
+    private String email;
+    private String password;
+    private String role; // USER, ADMIN
```
`role` 은 나중에 enum 값으로 관리해도 괜찮겠네요 🙂

woodying commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -26,4 +26,16 @@ public ErrorResponse handleValidationException(MethodArgumentNotValidException e
+    @ExceptionHandler(LoginFailedException.class)
+    @ResponseStatus(HttpStatus.FORBIDDEN)  // 403 Forbidden
+    public ErrorResponse handleLoginFailed(LoginFailedException ex) {
+        return new ErrorResponse(403, ex.getMessage());
```
403 -> `HttpStatus.FORBIDDEN.value()`
로 표현하면 더 좋을 것 같아요

woodying commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -26,4 +26,16 @@ public ErrorResponse handleValidationException(MethodArgumentNotValidException e
+    @ExceptionHandler(DuplicateEmailException.class)
+    @ResponseStatus(HttpStatus.CONFLICT)
+    public ErrorResponse handleDuplicateEmail(DuplicateEmailException ex) {
+        return new ErrorResponse(409, ex.getMessage());
```
409 -> `HttpStatus.CONFLICT.value()`
여기도 마찬가지 입니다!

woodying commented on src/main/java/gift/service/AuthMemberService.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public MemberResponseDto signUp(MemberRequestDto requestDto) {
+        // 이메일 중복 체크
+        if (memberRepository.findByEmail(requestDto.email()).isPresent()) {
+            throw new DuplicateEmailException(requestDto.email());
+        }
```
섬세한 처리 좋습니다 👍👍

woodying commented on src/main/java/gift/util/JwtUtil.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.Date;
+
+@Component
+public class JwtUtil {
```
진우님은 `JwtUtil` 을 왜 spring bean 으로 등록하는 결정을 하셨나요? 🙂
아주 잘해주셨는데요, `JwtUtil` 는 외부에서 의존성을 주입받는 것도 아니니까 필드와 함수 모두 static 으로 선언해서 사용해도 문제는 없어보여요. 그런데 진우님이 이렇게 판단하신 근거가 궁금합니다~! 🤔

jinwoojwa replied:
처음 스프링에 대해 배웠을 때, 스프링 기반 프로젝트에서는 모듈 간 결합도를 낮추고, 유지 보수를 용이하게 하기 위해 의존성 주입을 통해 클래스를 사용한다고 배웠습니다. 또한, 앞으로 미션이 진행되면서 요구사항에 따른 코드 변경이 어떻게 될 지 예측할 수 없어 `bean`으로 등록하는 방식을 사용했습니다.

<br>

질문을 통해 학습을 진행하면서 `Bean`으로 관리하는 방식과 `static` 방식 각각의 장단점에 대해 고민해보았습니다.

✅ Spring Bean으로 사용할 때

- 유연함, 확장성
- 테스트가 쉬움 (Mokito 사용 시 static 방식보다 편리하게 사용 가능)
- DI(의존성 주입)을 통해 클래스 간 결합도를 낮출 수 있음

✅ static 방식으로 사용할 때

- 간결함
- 반복 호출 시 성능이 더 나을 수 있음 (메모리에 한 번만 올라감)
- 상태가 없는 경우 static이 더 적합할 수 있음

woodying reviewed (changes requested):
진우님 안녕하세요 🙂

역시 진우님 PR 정리도 잘해주시고, 코드도 너무 깔끔하네요. 👍👍

추가 구현 계획 및 보충사항에 말씀주신 것처럼,
비밀번호를 평문으로 저장하면 문제가 있습니다. 🙂
개발자들이 유저들의 비밀번호를 조회해볼 수 있는 거니까요. 
그래서 비밀번호를 저장할때는 단방향 해시값으로 인코딩해서 저장합니다~
로그인 할 때도, 비밀번호의 해시값이 동일한 지 비교합니다. ☺️

진우님이라면 인증 필터와 관리자화면도 금방 깔끔하게 구현하실 것 같은데요~!
추가 구현 진행가능할까요? 🙏

제가 코멘트 드린 내용도 한 번 확인부탁드립니다.

jinwoojwa committed:
refactor: 예외 처리 코드에서 상태 코드를 HttpStatus 상수로 변경

jinwoojwa committed:
refactor: Member 엔티티 및 관련 코드에서 role 필드를 String -> enum 타입으로 변경

jinwoojwa committed:
chore: spring security 의존성 추가

jinwoojwa committed:
feat: 회원 비밀번호 암호화 기능 추가

jinwoojwa committed:
feat: Spring Security 인증 필터 구현

woodying reviewed (changes requested):
진우님 안녕하세요 😊

이번 미션 요구사항은 `spring security` 를 사용할 정도로 복잡한 요구사항은 없었던 것 같아요~!
간단한 요구사항에 비해 너무 무거운 라이브러리인 것 같습니다.
`JwtAuthenticationFilter` 를 잘 만들어두셔서, 이것만 활용해도 충분해 보여요🙂

`spring security` 의존성은 제거해주실 수 있을까요?

jinwoojwa committed:
refactor: Spring Security 제거 및 코드 리팩토링

jinwoojwa committed:
refactor: 클래스 및 dto, 인터페이스 이름 변경

- Rename : MemberController.java -> MemberApiController.java
- 회원에 대한 dto를 MemberResponseDto, TokenResponseDto로 분리
- Rename : AuthMemberService.java -> AuthenticationService.java
- 인터페이스, 클래스 역할을 명확히 하기 위해 이름 변경

jinwoojwa committed:
feat: 관리자 회원 관리 화면 및 관련 모달, JS 기능 추가

jinwoojwa committed:
feat: 관리자 회원 CRUD 기능 구현 및 서비스, 레포지토리 로직 추가

jinwoojwa commented:
피드백 해주신 사항들 다음과 같이 변경했습니다.

- 예외 처리 상태 코드를 HttpStatus enum을 활용하여 변경
- `role` 을 enum 으로 변경
- 인증 필터 구현(spring security 제거)
- 비밀번호 암호화(SHA-256 해시 함수 사용)
- Spring Security 코드 제거

**추가 구현**

- 관리자 회원 관리 페이지 추가

**질문**

미션을 진행할수록 클래스 파일이 점점 많아지면서 전체 구조가 한 눈에 들어오지 않는다는 생각을 가지게 되었습니다.

물론, controller, dto, service 등과 같이 클래스 파일들을 분류해서 넣어 놓고 있지만, 이름만 보고는 이게 어떤 엔티티에 관련된 서비스(혹은

레포지토리) 코드인지 일일이 파일을 열어봐야 하는 경우도 있었습니다. (클래스 네이밍 문제인 것 같기도 하지만, 전반적으로 패키지 구조가 더 직관적일 수 있지 않을까 고민하게 되었습니다)

❓ 혹시 적용해볼만한 패키지 구조나 개선 방안이 있는지 궁금합니다.

<br>

항상 꼼꼼한 피드백 감사드립니다. 조언해주시는 피드백 덕분에 코딩하면서 고민해야 할 지점들을 더 명확하게 인식하고, 구조적인 부분까지

도 한 번 더 되짚어보게 되는 것 같습니다.

woodying commented on src/main/java/gift/config/FilterConfig.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class FilterConfig {
```
👍

woodying commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -12,18 +12,30 @@ public class GlobalExceptionHandler {
     @ResponseStatus(HttpStatus.BAD_REQUEST)
     public ErrorResponse handleValidationException(MethodArgumentNotValidException ex) {
-        return new ErrorResponse(400, ex.getBindingResult().getFieldError().getDefaultMessage());
+        return new ErrorResponse(HttpStatus.BAD_REQUEST.value(), ex.getBindingResult().getFieldError().getDefaultMessage());
```
👍

woodying commented on src/main/java/gift/security/PasswordUtil.java:
```diff
@@ -0,0 +1,25 @@
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+
+public class PasswordUtil {
+
+    // SHA-256 해싱 + Base64 인코딩
+    public static String sha256(String input) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            byte[] hashBytes = md.digest(input.getBytes(StandardCharsets.UTF_8));
+            return Base64.getEncoder().encodeToString(hashBytes); // Base64로 문자열 변환
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // 사용 예시 (비교는 직접 문자열 비교)
+    public static boolean matches(String rawPassword, String hashedPassword) {
+        return sha256(rawPassword).equals(hashedPassword);
+    }
+}
```
인코더도 직접 만드셨네요! 👍

woodying reviewed (approved):
진우님 안녕하세요! 

역시 진우님 코드는 군더더기가 많이 없어서 보기 편합니다. 👍
더불어 아키텍처에 대한 고민까지 하고 계시군요~ 🙂

말씀주신대로 점점 구현사항이 많아지다보면 레이어별 하나의 패키지에 모든걸 담으면 보기 힘들어지는 경우가 있습니다. 지금 상황에서 써볼만한 구조로는 **계층형 아키텍처**를 유지하면서 domain 단위로 패키지를 만드는 방법이 있을 것 같습니다~ 🙂
예를 들어 이렇게요.

```
java/gift
    ├── common
    │   ├── auth
    │   ├── config
    │   └── exception
    │
    ├── member
    │   ├── controller
    │   ├── service
    │   ├── repository
    │   ├── entity
    │   └── dto
    │
    └── product
        ├── controller
        ├── service
        ├── repository
        ├── entity
        └── dto
```

기능도 잘 구현해주시고, 더 고쳐야할 부분은 없어보여서 merge 하겠습니다.
고생 많으셨어요! 👏
